### PR TITLE
Adapt batm to Smartcashj

### PR DESCRIPTION
**Pull request by pricky**
- Created two new classes
- Interfaces adapted
- Created some tests that covers ~70%
- gradle adapted for the maven dependency

### CODE REVIEW

**Comment by pricky** [comment]
Hi, 
Let me know if everything is in a right manner for you.
If you want that I modify something tell me.

Thank GB team

**Comment by rpanak-generalbytes** [comment]
Hi, there are a couple of problems with this approach.

1) we don't directly consume artifacts from JitPack. Ideally, we use artifacts from MavenCentral or JCenter, or in case of internal/forked libraries, we use our own repository. We use JitPack only so that contributors can actually build the project without accessing our internal repo. For the final build however, no JitPack-built artifacts are used. This could be solved by smartcash publishing on MavenCentral or JCenter.
2) we're experiencing memory-leak problems with bitcoinj and forks. Because of this we're reluctant to add another bitcoinj fork dependency. We'll have to discuss internally whether we'll allow this. If so, smartcashj would probably need to be based on a very recent bitcoinj version with the (known) leak fixes. Before we proceed we'd like to know whether it is possible for you to implement the functionality without smartcashj dependency.

**Comment by pricky** [comment]
> Hi, there are a couple of problems with this approach.
> 
> 1. we don't directly consume artifacts from JitPack. Ideally, we use artifacts from MavenCentral or JCenter, or in case of internal/forked libraries, we use our own repository. We use JitPack only so that contributors can actually build the project without accessing our internal repo. For the final build however, no JitPack-built artifacts are used. This could be solved by smartcash publishing on MavenCentral or JCenter.
> 2. we're experiencing memory-leak problems with bitcoinj and forks. Because of this we're reluctant to add another bitcoinj fork dependency. We'll have to discuss internally whether we'll allow this. If so, smartcashj would probably need to be based on a very recent bitcoinj version with the (known) leak fixes. Before we proceed we'd like to know whether it is possible for you to implement the functionality without smartcashj dependency.

Hi, 
We adapted the point 1.
The point 2 is more difficult to solve. I created some tests that show you what we will use. In this case the functions are light and you can test by yourself .

Let me know what we can do. We need this functionality asap as we ordered several machines... 
thanks

**Comment by rpanak-generalbytes** [comment]
Hi, the code seems not to be using the leaking parts of the library. 
However, there is a conflict in one of the files now. Can you please resolve it?

**Comment by pricky** [comment]
Hi, wich file is in conflict ? I can resolve ASAP. Let me know what is wrong.

Thank you

**Comment by rpanak-generalbytes** [comment]
You can see that just below the comments

**Comment by pricky** [comment]
Should I make an other pull request ?

**Comment by rpanak-generalbytes** [comment]
I think you messed up the conflict resolution. Take a look at your diff.

**Comment by pricky** [comment]
Hi, 
You're right I missed something. It is ok now ?

**Comment by rpanak-generalbytes** [comment]
Hi,

the diff still shows some changes that shouldn't be required, like the dependencySubstitutions block in Gradle file.

**Comment by pricky** [comment]
Hi,
I moved the content to the file dependencySubstitutions.txt. 
Is it okay for you ??

**Comment by rpanak-generalbytes** [comment]
- first of all, I don't think the changes are buildable now (you changed the dependencyChecksums.txt)
- take a look at the diff (https://github.com/GENERALBYTESCOM/bitrafael_public/pull/27/files). Are all of those changes really needed? I don't think so.
- you define a variable with smartcashj version 0.16.0, yet you don't use it anywhere (and use the version 0.17.2 instead)
- what is the reason for all those new dependency substitutions?
I won't look at the code until after these problems are resolved.

**Comment by pricky** [comment]
I restored the checksum.
I added only dependencysubstitutions. It will not build without that. 
The version is removed. It was created to respect your version naming. As we only use once the version, I removed that constant.

**Comment by rpanak-generalbytes** [comment]
Do you need the jitpack repo definition that you added?

**Comment by pricky** [comment]
No, I removed that. I forget the import. Thank you

**Comment by rpanak-generalbytes** [comment]
ok, looks much better now. I'd ask you to sort the dependency substitutions so that the same module substitutions are grouped together. I'll take a look at the code today or tomorrow.

**Comment by pricky** [comment]
It's done. 
it's perfect, I hope that it will be fine.

File: bitrafael-client/src/main/java/com/generalbytes/bitrafael/api/wallet/smart/WalletToolsSMART.java [hunk 1]
```diff
@@ -0,0 +1,265 @@
...
+        if (address.startsWith(optionalPrefix)) {
+            address = address.substring(optionalPrefix.length(), address.length());
+        }
+
+        if (!(address.startsWith("S"))) {
+            return false;
+        }
+
+        try {
+            Base58.decodeToBigInteger(address);
```
Comment (rpanak-generalbytes) [inline bitrafael-client/src/main/java/com/generalbytes/bitrafael/api/wallet/smart/WalletToolsSMART.java]: why is this needed?

File: bitrafael-client/src/test/java/com/generalbytes/bitrafael/api/wallet/smartcash/SmartcashjTest.java [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+
+import com.generalbytes.bitrafael.api.client.IClient;
+import com.generalbytes.bitrafael.api.wallet.IMasterPrivateKey;
+import com.generalbytes.bitrafael.api.wallet.IWalletTools;
+import com.generalbytes.bitrafael.api.wallet.smart.WalletToolsSMART;
+import org.junit.Before;
+import org.junit.Test;
+
+
+public class SmartcashjTest {
```
Comment (rpanak-generalbytes) [inline bitrafael-client/src/test/java/com/generalbytes/bitrafael/api/wallet/smartcash/SmartcashjTest.java]: Tests are insufficient. System.out.println doesn't test anything. Implement proper test cases.

**Comment by generalbytes** [comment]
I'm sorry we decided not to have smartcashj as a part of official release of bitrafael.
